src/hypofuzz/dashboard/dashboard.py

```diff
@@ -559,6 +549,37 @@ def send_nowait_from_anywhere(msg: ListenerEventT) -> None:
         )
         TESTS[fuzz_target.nodeid] = test
 
+        async def update_websockets() -> None:
+            # TODO: make this more granular, so we send incremental batches
+            # of reports as they're loaded, etc.
+            for websocket in websockets.copy():
+                await websocket.initial({test.nodeid: test})
+
+        trio.from_thread.run(update_websockets, trio_token=trio_token)
+
+
+async def run_dashboard(port: int, host: str) -> None:
+    send_channel, receive_channel = trio.open_memory_channel[ListenerEventT](math.inf)
+    trio_token = trio.lowlevel.current_trio_token()
+
+    def send_nowait_from_anywhere(msg: ListenerEventT) -> None:
+        # DirectoryBasedExampleDatabase sends events from a background thread (via watchdog),
+        # so we need to support sending from anywhere, i.e. whether or not the calling thread
+        # has any Trio state.  We can do that with the following branch:
+        try:
+            trio.lowlevel.current_task()
+        except RuntimeError:
+            trio.from_thread.run_sync(
+                send_channel.send_nowait, msg, trio_token=trio_token
+            )
+        else:
+            send_channel.send_nowait(msg)
+
+    db = get_db()
+    thread = threading.Thread(
+        target=_load_initial_state, kwargs={"trio_token": trio_token}
+    )
+    thread.start()
```

Comment:
Looking at this more closely, starting a thread which we never join is a bit of a code smell.

What if we instead had a load_initial_state async function which loaded _one_ test and then started listening for updates, which internally used `trio.to_thread.run_sync` for the database interaction?  Then we could just dump them all into the same nursery which serves the webapp, and we'd get nice cleanup for everything.

Reply:
so (pseudocode):
```
run_dashboard
    for test in test:
        nursery.start_soon(load_one_test)
    add_listener(listener)
```

with load_one_test setting some global state about when a test is currently being read (which our `listener` reads/respects), per your other comment.

(I just want to make sure you weren't intending each `load_one_test` to start its own listener; we should have one global listener per dashboard instance)

Reply:
yep, that's correct.

Then whatever is consuming our database-events channel will silently drop saves/deletes from tests which haven't finished loading yet, and process the rest in the ordinary way.